server/tests/classes/client.test.js

```diff
@@ -0,0 +1,347 @@
+      expect(c.connected).toBe(false);
+    });
+  });
+
+  describe('setPing', () => {
+    test.skip('should set ping', () => {
+      // TODO: What is this function for?
+    });
+  });
+
+  describe('setFlight', () => {
+    test('should set the flight', () => {
+      const c = new Client();
+      c.setFlight('nonExistingFlightId');
+      expect(c.flightId).toBe('nonExistingFlightId');
+    });
+
+    test('should reset station and hypercard', () => {
+      const c = new Client({station: 'fakeStation', hypercard: 'fakeHypercard'});
+      c.setFlight('nonExistingFlightId');
+      expect(c.station).toBe(null);
+      expect(c.hypercard).toBe(null);
+    });
+
+    test.skip('should set flight if flightId is found in App', () => {
+      // TODO: How to add a flight?
+    });
+  })
+
+  describe('setSimulator', () => {
+    test('should set the simulatorId', () => {
+      const c = new Client();
+      c.setSimulator('test_simulator_id');
+      expect(c.simulatorId).toBe('test_simulator_id');
+    });
+
+    test('should reset station', () => {
+      const c = new Client({ station: 'test_station' });
+      c.setSimulator('test_simulator_id');
+      expect(c.station).toBeNull();
+    });
+  })
+
+  describe('setStation', () => {
+    test('should set the station', () => {
+      const c = new Client();
+      c.setStation('test_station');
+      expect(c.station).toBe('test_station');
+    });
+  });
+
+  describe('login/logout', () => {
+    test('should set loginName and loginState', () => {
+      const c = new Client();
+
+      c.login('kermit');
+      expect(c.loginName).toBe('kermit');
+      expect(c.loginState).toBe('login');
+
+      c.logout();
+      expect(c.loginName).toBeNull();
+      expect(c.loginState).toBe('logout');
+    });
+  });
+
+  describe('setTraining', () => {
+    test('should set the training', () => {
+      const c = new Client();
+      expect(c.training).toBe(false);
+      c.setTraining(true);
+      expect(c.training).toBe(true);
+    });
+  });
+
+  describe('setHypercard', () => {
+    test('should set the hypercard', () => {
+      const c = new Client();
+      c.setHypercard('test_card');
+      expect(c.hypercard).toBe('test_card');
+      c.setHypercard();
+      expect(c.hypercard).toBeNull();
+    });
+  });
+
+  describe('setMovie', () => {
+    test('should set a movie', () => {
+      const c = new Client();
+      c.setMovie('not_a_real_movie');
+      expect(c.movie).toBe('not_a_real_movie');
+      expect(c.offlineState).toBe('movie');
+    });
+  });
+
+  describe('setOfflineState', () => {
+    test('should set the offline state and clear the movie', () => {
+      const c = new Client();
+      c.setMovie('not_a_real_movie');
+      c.setOfflineState('blackout');
+      expect(c.movie).toBeNull();
+      expect(c.offlineState).toBe('blackout');
+    });
+  });
+
+  describe('addCache', () => {
+    test.skip('should work', () => {
+      // TODO: What does this do?
```

Comment:
The Client object is a reference to a connected client. Caches allow the server to tell the client large files that it should preload, such as movies, images, or sound effects. All it does is throw the URI of the file in an array and push that to the client with a subscription publish. The client then uses a `<link rel="preload">` to preload the asset

Reply:
I will add some tests for this but I can't get to right this moment.

Reply:
It's fine. You can leave the `// TODO` statements and we can get to this later.